File: crates/nil/src/commands/player.ts
import { invoke } from '@tauri-apps/api/core';

export function getPlayer(id: PlayerId) {
  return invoke<Player>('get_player', { id });
}

export async function getPlayerVillages(id: PlayerId) {
  return invoke<readonly Coord[]>('get_player_villages', { id });
}

export function getPlayers() {
  return invoke<readonly Player[]>('get_players');
}

export function removePlayer(id: PlayerId) {
  return invoke<null>('remove_player', { id });
}

export function spawnPlayer(options: PlayerOptions) {
  return invoke<null>('spawn_player', { options });
}

File: crates/nil/src-tauri/src/command/player.rs
use crate::error::CResult;
use crate::manager::ManagerExt;
use nil_core::{Coord, Player, PlayerId, PlayerOptions};
use tauri::AppHandle;

#[tauri::command]
pub async fn get_player(app: AppHandle, id: PlayerId) -> CResult<Player> {
  app
    .client(async |cl| cl.player(id).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_villages(app: AppHandle, id: PlayerId) -> CResult<Vec<Coord>> {
  app
    .client(async |cl| cl.villages_of(id).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_players(app: AppHandle) -> CResult<Vec<Player>> {
  app
    .client(async |cl| cl.players().await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn remove_player(app: AppHandle, id: PlayerId) -> CResult<()> {
  app
    .client(async |cl| cl.remove_player(id).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn spawn_player(app: AppHandle, options: PlayerOptions) -> CResult<()> {
  app
    .client(async |cl| cl.spawn_player(options).await)
    .await?
    .map_err(Into::into)
}
